File: mult.py
import random

print("Enter *+* to quit")

def mult_prac():
	while True:
		num_in = input("Number to practice: ")
		if num_in == "*+*":
			return
		try:
			int_in = int(num_in)
			break
		except ValueError:
			print("Please enter a number.")



	#create list for wrong numbers
	wrong = []

	#create list of correct and incorrect messages
	correct = ["Good job!", "Right!", "Yep!", "Nailed it.", "Spot on!", "Bingo!", "Crushed it.", "Totally!", "Perfecto!", "You are killing it.", "Genius!", "Fantastic!", "You rock!", "Awesome!", "Niiiiiiiice!", "Yaaaasssss!", "So. Dang. Smart."]
	incorrect = ["Nope", "Sorry, that's not it", "Maybe next time", "No. Focus!", "Hmmm, no.", "Uuuuuuh, no.", "Ew, no!.", "Nasty! No way."]

	# Create list of numbers 1-12 named num_list
	num_list = list(range(1, 13))


	#Is list empty? no:
	while len(num_list) > 0:

		# Get random number from num_list
		test = num_list.pop(random.randrange(len(num_list)))
		num_test = int(test)
		print()
		#ask user math problem, get answer
		while True:
			eval = input(f"What is {test} x {num_in}? ")

			if eval == "*+*":
				return
			try:
				int_eval = int(eval)
				break
			except ValueError:
				print("Please enter a number.")
		# is eval correct?
		if int_eval == num_test*int_in:
			print(random.choice(correct))
		else:
			print(random.choice(incorrect))
			wrong.append(num_test)

	if len(wrong) == 0:
			print()
			print("Wow, 100%!!\n")
			return
	else:
		num_wrong = len(wrong)
		while len(wrong) > 0:
			print()
			test = wrong.pop(random.randrange(len(wrong)))
			eval = input(f"What is {test} x {int_in}? ")
			int_eval = int(eval)
			if int_eval == test*int_in:
				print(random.choice(correct))
			else:
				print(random.choice(incorrect))
				wrong.append(test)
				num_wrong = num_wrong + 1
		print()
		print(f"You got them all!\nYou missed {num_wrong}.\n")
		return

File: mult.js
let wrong = [];
let numToPractice = 0;
let numList = [];
let currentTest = 0;
let correctAnswers = 0;
let initialMissed = 0;
let lastMessage = ""; // Track the last feedback message
const correctMessages = ["Good job!", "Right!", "Yep!", "Nailed it.", "Spot on!", "Bingo!", "Crushed it.", "Totally!", "Perfecto!", "You are killing it.", "Genius!", "Fantastic!", "You rock!", "Awesome!", "Niiiiiiiice!", "Yaaaasssss!", "So. Dang. Smart."];
const incorrectMessages = ["Nope", "Sorry, that's not it", "Maybe next time", "No. Focus!", "Hmmm, no.", "Uuuuuuh, no.", "Ew, no!.", "Nasty! No way."];

// Set focus on the input when the page loads
window.onload = function() {
    document.getElementById("numInput").focus();
};

// Start practice function
function startPractice() {
    const input = document.getElementById("numInput").value;
    document.getElementById("feedback").textContent = ""; // Clear feedback
    document.getElementById("summary").textContent = ""; // Clear summary

    const intIn = parseInt(input);
    if (isNaN(intIn)) {
        document.getElementById("feedback").textContent = "Please enter a valid number.";
        return;
    }

    numToPractice = intIn;
    numList = Array.from({length: 12}, (_, i) => i + 1); // list from 1 to 12
    correctAnswers = 0;
    initialMissed = 0;
    wrong = [];

    document.getElementById("answerInput").style.display = "inline";
    document.getElementById("submitAnswerBtn").style.display = "inline";
    document.getElementById("answerInput").focus(); // Focus the input field after the practice starts
    askQuestion();
}

// Ask the next question
function askQuestion() {
    if (numList.length > 0) {
        const randomIndex = Math.floor(Math.random() * numList.length);
        currentTest = numList.splice(randomIndex, 1)[0]; // Get random number from list
        document.getElementById("question").textContent = `${currentTest} x ${numToPractice} =`;
        document.getElementById("answerInput").focus(); // Ensure the cursor is in the answer input field
    } else if (wrong.length > 0) {
        currentTest = wrong.pop();
        document.getElementById("question").textContent = `Retry: ${currentTest} x ${numToPractice} =`;
        document.getElementById("answerInput").focus();
    } else {
        endPractice();
    }
}

// Handle submission of answers
function submitAnswer() {
    const answer = parseInt(document.getElementById("answerInput").value);
    
    if (isNaN(answer)) {
        document.getElementById("feedback").textContent = "Please enter a valid number.";
        return;
    }

    const feedbackElement = document.getElementById("feedback");

    if (answer === currentTest * numToPractice) {
        const message = randomMessage(correctMessages);
        feedbackElement.textContent = message;
        feedbackElement.style.color = "#32CD32";  // Brighter lime green for correct
        feedbackElement.style.fontWeight = "bold"; // Bold font for correct
        lastMessage = message;
        correctAnswers++;
    } else {
        const message = randomMessage(incorrectMessages);
        feedbackElement.textContent = message;
        feedbackElement.style.color = "red";   // Incorrect answer in red
        feedbackElement.style.fontWeight = "bold"; // Bold font for incorrect
        lastMessage = message;
        wrong.push(currentTest);
        if (numList.length > 0) initialMissed++; // Only count initial missed
    }

    document.getElementById("answerInput").value = ""; // Clear input for next question
    askQuestion();
}

// Get random message ensuring it's not repeated
function randomMessage(arr) {
    let message;
    do {
        message = arr[Math.floor(Math.random() * arr.length)];
    } while (message === lastMessage); // Prevent repeating the same message twice in a row
    return message;
}

// End practice session and show summary
function endPractice() {
    document.getElementById("question").textContent = "";
    document.getElementById("answerInput").style.display = "none";
    document.getElementById("submitAnswerBtn").style.display = "none";

    if (initialMissed === 0) {
        document.getElementById("summary").textContent = `Wow, 100%! You got them all right on the first try!`;
    } else {
        document.getElementById("summary").textContent = `You got them all! This time you missed ${initialMissed}.`;
    }
}

// Allow the "Enter" key to start practice
document.getElementById("numInput").addEventListener("keyup", function(event) {
    if (event.key === "Enter") {
        startPractice();
    }
});

// Allow the "Enter" key to submit the answer
document.getElementById("answerInput").addEventListener("keyup", function(event) {
    if (event.key === "Enter") {
        submitAnswer();
    }
});